File: leetcode/chain/234.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def reverse(self, head: ListNode) -> ListNode:
        ret = None
        while head:
            tmp = head.next
            head.next = ret
            ret = head
            head = tmp
        return ret

    def isPalindrome(self, head: ListNode) -> bool:
        '''快慢指针求中间,反转右半,进行顺序判断'''
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        m = slow.next
        slow.next = None

        m = self.reverse(m)
        while head and m:
            if head.val != m.val:
                return False
            head = head.next
            m = m.next
        return True

File: leetcode/chain/234.cpp
#include <iostream>

#include "list_node.h"

class Solution {
 public:
  ListNode *reverse(ListNode *node) {
    ListNode * last = nullptr;
    while (node) {
      ListNode * tmp = node->next;
      node->next = last;

      last = node;
      node = tmp;
    }
    return last;
  }

  bool is_palindrome(ListNode* node) {
    ListNode *slow = node, *fast = node->next;
    while (fast && fast->next) {
      fast = fast->next->next;
      slow = slow->next;
    }

    ListNode* second = slow->next;
    slow->next = nullptr;

    ListNode* reverse_node = reverse(second);
    while (node && reverse_node) {
      if (node->val != reverse_node->val) {
        return false;
      }

      node = node->next;
      reverse_node = reverse_node->next;
    }
    return true;
  }
};

int main() {
  ListNode* node = new ListNode(1);
  node->append(new ListNode(2));
  node->print();

  Solution solution = Solution();
  std::cout << solution.is_palindrome(node) << std::endl;
}